// miew-react/dist/index.js
/** 3D Molecular Viewer v1.0.0
Copyright (c) 2015-2022 EPAM Systems, Inc. */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var jsxRuntime = require('@emotion/react/jsx-runtime');
var react = require('react');
var reactRedux = require('react-redux');
var miew = require('miew');
var useResizeObserver = require('use-resize-observer');
var react$1 = require('@emotion/react');
var styles = require('@mui/material/styles');
var material = require('@mui/material');
var lodash = require('lodash');
var createSagaMiddleware = require('redux-saga');
var toolkit = require('@reduxjs/toolkit');
var _regeneratorRuntime = require('@babel/runtime/regenerator');
var effects = require('redux-saga/effects');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var useResizeObserver__default = /*#__PURE__*/_interopDefaultLegacy(useResizeObserver);
var createSagaMiddleware__default = /*#__PURE__*/_interopDefaultLegacy(createSagaMiddleware);
var _regeneratorRuntime__default = /*#__PURE__*/_interopDefaultLegacy(_regeneratorRuntime);

var initialState = {
  isInitialized: false,
  data: undefined,
  error: undefined
};
var InitSlice = toolkit.createSlice({
  name: 'init',
  initialState: initialState,
  reducers: {
    INIT: function INIT(state) {
      state.isInitialized = false;
      state.data = undefined;
      state.error = undefined;
    },
    INIT_SUCCESS: function INIT_SUCCESS(state, action) {
      state.isInitialized = true;
      state.data = action.payload;
      state.error = undefined;
    },
    INIT_FAIL: function INIT_FAIL(state, action) {
      state.isInitialized = false;
      state.data = undefined;
      state.error = action.payload;
    }
  }
});
var _InitSlice$actions = InitSlice.actions,
    INIT = _InitSlice$actions.INIT,
    INIT_SUCCESS = _InitSlice$actions.INIT_SUCCESS,
    INIT_FAIL = _InitSlice$actions.INIT_FAIL;
var initReducer = InitSlice.reducer;

var _marked$1 = _regeneratorRuntime__default["default"].mark(initApp),
    _marked2 = _regeneratorRuntime__default["default"].mark(watchInitApp);
var fetchInitData = function fetchInitData(url) {
  return new Promise(function (resolve) {
    return setTimeout(function () {
      return resolve("data fetched from ".concat(url));
    }, 2000);
  });
};
var TEST_URL = 'some/url';
function initApp() {
  var data;
  return _regeneratorRuntime__default["default"].wrap(function initApp$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.prev = 0;
          _context.next = 3;
          return effects.call(fetchInitData, TEST_URL);
        case 3:
          data = _context.sent;
          _context.next = 6;
          return effects.put(INIT_SUCCESS(data));
        case 6:
          _context.next = 12;
          break;
        case 8:
          _context.prev = 8;
          _context.t0 = _context["catch"](0);
          _context.next = 12;
          return effects.put(INIT_FAIL('SAMPLE ERROR'));
        case 12:
        case "end":
          return _context.stop();
      }
    }
  }, _marked$1, null, [[0, 8]]);
}
function watchInitApp() {
  return _regeneratorRuntime__default["default"].wrap(function watchInitApp$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return effects.takeEvery(INIT, initApp);
        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2);
}

var _marked = _regeneratorRuntime__default["default"].mark(rootSaga);
function rootSaga() {
  return _regeneratorRuntime__default["default"].wrap(function rootSaga$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return effects.spawn(watchInitApp);
        case 2:
        case "end":
          return _context.stop();
      }
    }
  }, _marked);
}

var sagaMiddleware = createSagaMiddleware__default["default"]();
var store = toolkit.configureStore({
  reducer: {
    init: initReducer
  },
  middleware: function middleware(getDefaultMiddleware) {
    return getDefaultMiddleware().concat(sagaMiddleware);
  }
});
sagaMiddleware.run(rootSaga);

var defaultTheme = {
  palette: {
    accent: {
      main: '#2E6DA4',
      dark: '#204D74',
      light: '#337AB7'
    },
    primary: {
      main: '#DDD',
      light: '#FFF',
      dark: '#DDD'
    },
    secondary: {
      main: '#333',
      light: '#C0C0C0',
      dark: '#383838'
    }
  },
  typography: {
    fontSize: 16,
    fontFamily: 'sans-serif'
  }
};

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var MEDIA_SIZES = {
  smallWidth: 800,
  smallHeight: 400
};
var muiTheme = styles.createTheme();
var Viewer = function Viewer(_ref) {
  var onInit = _ref.onInit,
      options = _ref.options,
      theme = _ref.theme;
  var viewerTheme = theme ? lodash.merge(defaultTheme, theme) : defaultTheme;
  var ref = react.useRef(null);
  var _useResizeObserver = useResizeObserver__default["default"]({
    ref: ref
  }),
      width = _useResizeObserver.width,
      height = _useResizeObserver.height;
  var isSizeSmall = height && height <= MEDIA_SIZES.smallHeight || width && width <= MEDIA_SIZES.smallWidth;
  var viewerStyle = function viewerStyle(theme) {
    var palette = theme.miew.palette;
    return {
      backgroundColor: isSizeSmall ? palette.accent.main : palette.primary.main,
      height: '100%',
      width: '100%',
      '& > .miew-canvas': {
        height: '100%',
        width: '100%'
      }
    };
  };
  react.useLayoutEffect(function () {
    var miew$1 = new miew.Miew(_objectSpread({
      container: ref === null || ref === void 0 ? void 0 : ref.current
    }, options));
    if (miew$1.init()) miew$1.run();
    if (typeof onInit === 'function') onInit(miew$1);
  }, [options, onInit]);
  return jsxRuntime.jsx(reactRedux.Provider, {
    store: store,
    children: jsxRuntime.jsxs(react$1.ThemeProvider, {
      theme: lodash.merge(muiTheme, {
        miew: viewerTheme
      }),
      children: [jsxRuntime.jsx(material.CssBaseline, {}, void 0), jsxRuntime.jsx("div", {
        ref: ref,
        css: viewerStyle
      }, void 0)]
    }, void 0)
  }, void 0);
};

exports["default"] = Viewer;
